data/dataset_utils.py

```
@@ -181,6 +187,262 @@ def materialize_data(ds: datasets.IterableDataset) -> List[str]:
+
+    if path.endswith(".parquet"):
+        pf = pq.ParquetFile(path)
+        result = []
+        remaining = take
+        skip = local_offset
+        for batch in pf.iter_batches(batch_size=4096, use_pandas_metadata=True):
+            n = batch.num_rows
+
+            if skip >= n:
+                skip -= n
+                continue
+
+            start = skip
+            end = start + min(remaining, n - start)
+            batch_slice = batch.slice(start, end - start)
+            df = batch_slice.to_pandas()
+
+            if "text" in df.columns:
+                df = df[df["text"].notnull()]
+                df["text"] = df["text"].astype(str)
+                result.extend(df.to_dict(orient="records"))
+
+            remaining -= (end - start)
+            skip = 0
+            if remaining <= 0:
+                break
+        return result
+
+    result = []
+    with open_compressed_file(path) as f:
+        for line in islice(f, local_offset, local_offset + take):
+            line = line.strip()
+            if not line:
+                continue
+            try:
+                obj = json.loads(line)
+            except json.JSONDecodeError:
+                obj = {"text": line}
+            result.append(obj)
+    return result
+
+
+def read_sharded_across_files(files: List[str],
+                              world_size: int,
+                              global_rank: int,
+                              chunk_size: int,
+                              offset: int) -> List[Dict]:
+    """Read data sharded across multiple files for distributed training.
+    Args:
+        files: List of file paths to read from
+        world_size: Total number of shards
+        global_rank: Rank of current shard
+        chunk_size: Size of chunk to read
+        offset: Offset within the shard
+        
+    Returns:
+        List[Dict]: List of data records from the assigned shard
+    """
+    total_len = sum(get_file_len(f) for f in files)
```

Getting lengths for all files in a big dataset (e.g., C4-en has over 1000 chunks  https://huggingface.co/datasets/allenai/c4/tree/main/en) would be manageable only with Parquet datasets having `num_rows` metadata attribute. Otherwise, you'd need to decompress and read each of the thousands of files, at least at the initial pass before the first dataset construction. 

This would involve decompressing and reading up to terabytes of data and cause a very long latency (delay) before the start of the first epoch. 

If we want exact counting of rows, let's do it the following way: read files one at a time until their cumulative row count exceeds desired `offset`. That way, for each new epoch, we have to read at most one data file. And if we resume training, we still have to read and get lengths from a fraction of dataset files, rather than all of them.

---

Due to the absence of the total dataset length, each of the `global_rank` in `world_size` should have its own independent subset of files to iterate. This would make the algorithm much simpler, but it implies that the number of files should be at least equal to the number of nodes. Please raise an error if the condition is not met.